Upgrading from CAPI v0.3.14 to v0.3.22 can result in KCP NPE when a KCP upgrade is initiated without the KCP resource being updated

**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

Upgrade CAPI controllers from v0.3.14 to v0.3.22 . KCP rollout is triggered (presumably because a diff condition is reached) with no update to the KCP resource.

KCP controller starts crash looping with:

```
kubeadm-control-plane-controller-manager-b997596c9-rp54r -n capi-kubeadm-control-planrol-plane-controller-manager-b997596c9-rp54r -n capi-kubeadm-cone-system -c mtrol-plane-system -c manager-b997596c9-rp54r -n capi-kubeadm-control-plane-system -c manager-b997596c9-rp54r -n capi-kubeadm-control-plane-system -c manag
I0823 08:00:18.816115       1 listener.go:44] controller-runtime/metrics "msg"="metrics server is starting to listen"  "addr"="127.0.0.1:8080" -c manag
I0823 08:00:18.816829       1 main.go:147] setup "msg"="starting manager"  "version"=""
I0823 08:00:18.817421       1 leaderelection.go:242] attempting to acquire leader lease  capi-kubeadm-control-plane-system/kubeadm-control-plane-manager-leader-election-capi...
I0823 08:00:18.817477       1 internal.go:356] controller-runtime/manager "msg"="starting metrics server"  "path"="/metrics"
I0823 08:00:34.193122       1 leaderelection.go:252] successfully acquired lease capi-kubeadm-control-plane-system/kubeadm-control-plane-manager-leader-election-capi
I0823 08:00:34.200093       1 controller.go:155] controller-runtime/controller "msg"="Starting EventSource" "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"version":"","infrastructureTemplate":{},"kubeadmConfigSpec":{}},"status":{"initialized":false,"ready":false}}}
I0823 08:00:34.304736       1 controller.go:155] controller-runtime/controller "msg"="Starting EventSource" "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"clusterName":"","bootstrap":{},"infrastructureRef":{}},"status":{"bootstrapReady":false,"infrastructureReady":false}}}
I0823 08:00:34.405406       1 controller.go:155] controller-runtime/controller "msg"="Starting EventSource" "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"controlPlaneEndpoint":{"host":"","port":0}},"status":{"infrastructureReady":false,"controlPlaneInitialized":false}}}
I0823 08:00:34.505960       1 controller.go:162] controller-runtime/controller "msg"="Starting Controller" "controller"="kubeadmcontrolplane"
I0823 08:00:34.506227       1 controller.go:189] controller-runtime/controller "msg"="Starting workers" "controller"="kubeadmcontrolplane" "worker count"=10
I0823 08:00:34.519273       1 controller.go:246] controllers/KubeadmControlPlane "msg"="Reconcile KubeadmControlPlane" "cluster"="ndc-k8s-0" "kubeadmControlPlane"="ndc-k8s-master-0-control-plane" "namespace"="ndc-k8s-0"
I0823 08:00:34.519581       1 controller.go:246] controllers/KubeadmControlPlane "msg"="Reconcile KubeadmControlPlane" "cluster"="ndc-mgmt-0" "kubeadmControlPlane"="ndc-mgmt-master-0-control-plane" "namespace"="tkg-system"
I0823 08:00:34.520370       1 controller.go:246] controllers/KubeadmControlPlane "msg"="Reconcile KubeadmControlPlane" "cluster"="wc1" "kubeadmControlPlane"="wc1-m-control-plane" "namespace"="wc1"
I0823 08:00:36.017315       1 controller.go:331] controllers/KubeadmControlPlane "msg"="Rolling out Control Plane machines" "cluster"="ndc-mgmt-0" "kubeadmControlPlane"="ndc-mgmt-master-0-control-plane" "namespace"="tkg-system" "needRollout"=["ndc-mgmt-master-0-control-plane-77nsp"]
E0823 08:00:37.053767       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 337 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x183ed20, 0x278a1c0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/runtime/runtime.go:48 +0x89
panic(0x183ed20, 0x278a1c0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).upgradeControlPlane(0xc0000e2a10, 0x1c66e80, 0xc000054020, 0xc000871d40, 0xc0008aa900, 0xc0004f20a0, 0xc0008beb10, 0x7, 0x1aa31e6, 0x3f, ...)
	/workspace/controlplane/kubeadm/controllers/upgrade.go:103 +0x658
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).reconcile(0xc0000e2a10, 0x1c66e80, 0xc000054020, 0xc000871d40, 0xc0008aa900, 0x0, 0x0, 0x0, 0xc0007229d0)
	/workspace/controlplane/kubeadm/controllers/controller.go:333 +0xfb3
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).Reconcile(0xc0000e2a10, 0xc0007229d0, 0xa, 0xc00057ec60, 0x1f, 0x0, 0x0, 0x0, 0x0)
	/workspace/controlplane/kubeadm/controllers/controller.go:212 +0x726
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0007114a0, 0x18a4320, 0xc00056edc0, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:255 +0x166
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0007114a0, 0xc00042bf00)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:231 +0xb0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0007114a0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:210 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000535050)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:152 +0x5f
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000535050, 0x3b9aca00, 0x0, 0x1, 0xc000043d40)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:153 +0x105
k8s.io/apimachinery/pkg/util/wait.Until(0xc000535050, 0x3b9aca00, 0xc000043d40)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:192 +0x435
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x16da258]

goroutine 337 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x183ed20, 0x278a1c0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).upgradeControlPlane(0xc0000e2a10, 0x1c66e80, 0xc000054020, 0xc000871d40, 0xc0008aa900, 0xc0004f20a0, 0xc0008beb10, 0x7, 0x1aa31e6, 0x3f, ...)
	/workspace/controlplane/kubeadm/controllers/upgrade.go:103 +0x658
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).reconcile(0xc0000e2a10, 0x1c66e80, 0xc000054020, 0xc000871d40, 0xc0008aa900, 0x0, 0x0, 0x0, 0xc0007229d0)
	/workspace/controlplane/kubeadm/controllers/controller.go:333 +0xfb3
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).Reconcile(0xc0000e2a10, 0xc0007229d0, 0xa, 0xc00057ec60, 0x1f, 0x0, 0x0, 0x0, 0x0)
	/workspace/controlplane/kubeadm/controllers/controller.go:212 +0x726
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0007114a0, 0x18a4320, 0xc00056edc0, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:255 +0x166
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0007114a0, 0xc00042bf00)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:231 +0xb0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0007114a0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:210 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000535050)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:152 +0x5f
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000535050, 0x3b9aca00, 0x0, 0x1, 0xc000043d40)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:153 +0x105
k8s.io/apimachinery/pkg/util/wait.Until(0xc000535050, 0x3b9aca00, 0xc000043d40)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:192 +0x435
```

**What did you expect to happen:**

No NPE, KCP rollout succeeds, or doesn't happen.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Cluster-api version: v0.3.14 --> v0.3.22
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

--- a/controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go
+++ b/controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go
@@ -79,6 +79,7 @@ type KubeadmControlPlaneSpec struct {
 	// The RolloutStrategy to use to replace control plane machines with
 	// new ones.
 	// +optional
+	// +kubebuilder:default={type: "RollingUpdate", rollingUpdate: {maxSurge: 1}}
 	RolloutStrategy *RolloutStrategy `json:"rolloutStrategy,omitempty"`
 }
 

--- a/controlplane/kubeadm/config/crd/bases/controlplane.cluster.x-k8s.io_kubeadmcontrolplanes.yaml
+++ b/controlplane/kubeadm/config/crd/bases/controlplane.cluster.x-k8s.io_kubeadmcontrolplanes.yaml
@@ -2226,6 +2226,10 @@ spec:
                 format: date-time
                 type: string
               rolloutStrategy:
+                default:
+                  rollingUpdate:
+                    maxSurge: 1
+                  type: RollingUpdate
                 description: The RolloutStrategy to use to replace control plane machines
                   with new ones.
                 properties:

--- a/controlplane/kubeadm/config/crd/bases/controlplane.cluster.x-k8s.io_kubeadmcontrolplanetemplates.yaml
+++ b/controlplane/kubeadm/config/crd/bases/controlplane.cluster.x-k8s.io_kubeadmcontrolplanetemplates.yaml
@@ -1125,6 +1125,10 @@ spec:
                         format: date-time
                         type: string
                       rolloutStrategy:
+                        default:
+                          rollingUpdate:
+                            maxSurge: 1
+                          type: RollingUpdate
                         description: The RolloutStrategy to use to replace control
                           plane machines with new ones.
                         properties:

